ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java

```diff
@@ -172,31 +174,54 @@ public ConsistencyCheckerTool(ClusterMap clusterMap, StoreKeyFactory storeKeyFac
    * @return {@code true} if no real inconsistent blobs. {@code false}
    * @throws Exception
    */
-  public boolean checkConsistency(File[] replicas) throws Exception {
-    Map<StoreKey, ReplicationStatus> blobIdToStatusMap = getBlobStatusByReplica(replicas);
-    return checkConsistency(blobIdToStatusMap, replicas.length).size() == 0;
+  public Pair<Boolean, Map<File, DumpIndexTool.IndexProcessingResults>> checkConsistency(File[] replicas)
+      throws Exception {
+    Pair<Boolean, Map<File, DumpIndexTool.IndexProcessingResults>> resultsByReplica =
+        getIndexProcessingResults(replicas);
+    boolean success = resultsByReplica.getFirst();
+    if (success) {
+      Map<StoreKey, ReplicationStatus> blobIdToStatusMap =
+          getBlobStatusByReplica(replicas, resultsByReplica.getSecond());
+      success = checkConsistency(blobIdToStatusMap, replicas.length).size() == 0;
+    }
+    return new Pair<>(success, resultsByReplica.getSecond());
+  }
+
+  /**
+   * Processes the indexes of each of the replicas and returns the results.
+   * @param replicas the replicas to process indexes for.
+   * @return a {@link Pair} whose first indicates whether all results were sane and whose second contains the map of
+   * individual results by replica.
+   * @throws Exception if there is any error in processing the indexes.
+   */
+  private Pair<Boolean, Map<File, DumpIndexTool.IndexProcessingResults>> getIndexProcessingResults(File[] replicas)
+      throws Exception {
+    long currentTimeMs = time.milliseconds();
+    Map<File, DumpIndexTool.IndexProcessingResults> results = new HashMap<>();
+    boolean sane = true;
+    for (File replica : replicas) {
+      logger.info("Processing segment files for replica {} ", replica);
+      DumpIndexTool.IndexProcessingResults result =
+          dumpIndexTool.processIndex(replica, filterSet, currentTimeMs, throttler);
```

Comment:
what does `processIndex()` involve? (no javadocs for that method).

Reply:
Unfortunately many of the functions in the older tools don't have docs. In a nutshell, process index goes through all index files and
1) Determines if there are any duplicate puts/deletes/out of order entries
2) The state of a key in the index (valid/deleted/expired)
3) If some key is in a "recent" index segment